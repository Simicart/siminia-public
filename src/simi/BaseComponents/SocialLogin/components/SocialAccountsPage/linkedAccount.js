import React from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { bool, func, shape, string } from 'prop-types';

import defaultClasses from './linkedAccount.module.css';
import Button from '@magento/venia-ui/lib/components/Button';
import { FormattedMessage } from 'react-intl';
import SocialIcon from '../SocialIcon';

const LinkedAccount = props => {
  const { type, name, handleUnlink, isDisabled } = props;
  const classes = mergeClasses(defaultClasses, props.classes);

  return (
    <li className={classes.root}>
      <SocialIcon type={type} classes={{ root: classes.icon }} />
      <span className={classes.name}>{name}</span>
      <Button
        type="button"
        priority="high"
        disabled={isDisabled}
        onClick={() => handleUnlink(type)}
      >
        <FormattedMessage
          id={'amSocialLogin.unlink'}
          defaultMessage={'Unlink'}
        />
      </Button>
    </li>
  );
};

LinkedAccount.propTypes = {
  classes: shape({
    root: string
  }),
  type: string,
  name: string,
  handleUnlink: func,
  isDisabled: bool
};

LinkedAccount.defaultProps = {
  name: ''
};

export default LinkedAccount;
