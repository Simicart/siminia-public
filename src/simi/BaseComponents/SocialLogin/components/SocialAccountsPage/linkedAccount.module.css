.root {
  display: flex;
  align-items: center;
  min-width: 0;
  background: rgb(var(--venia-global-color-gray-100));
  margin-bottom: 0.5rem;
  padding: 0.5rem;
}

.icon {
  composes: root from '../SocialIcon/socialIcon.css';
  margin: 0 1rem 0 0;
  width: 2.5rem;
  height: 2.5rem;
  min-width: 2.5rem;
}

.name {
  max-width: 100%;
  margin-right: auto;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.root > button {
  min-width: 7rem;
}
