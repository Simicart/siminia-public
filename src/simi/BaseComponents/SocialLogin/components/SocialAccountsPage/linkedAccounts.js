import React from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { shape, string, array, func, bool } from 'prop-types';
import { FormattedMessage } from 'react-intl';
import LinkedAccount from './linkedAccount';
import defaultClasses from './socialAccountsPage.module.css';

const LinkedAccounts = props => {
  const { accounts, handleUnlink, isDisabled } = props;
  const classes = mergeClasses(defaultClasses, props.classes);

  if (!Array.isArray(accounts) || !accounts.length) {
    return null;
  }

  const list = accounts.map(acc => (
    <LinkedAccount
      key={acc.type}
      isDisabled={isDisabled}
      handleUnlink={handleUnlink}
      {...acc}
    />
  ));

  return (
    <div className={classes.accountsBlock}>
      <div className={classes.accountsHeading}>
        <FormattedMessage
          id={'amSocialLogin.linkedAccounts'}
          defaultMessage={'Linked accounts'}
        />
      </div>

      <ul className={classes.accountsList}>{list}</ul>
    </div>
  );
};

LinkedAccounts.propTypes = {
  classes: shape({
    root: string
  }),
  accounts: array,
  handleUnlink: func,
  isDisabled: bool
};

export default LinkedAccounts;
