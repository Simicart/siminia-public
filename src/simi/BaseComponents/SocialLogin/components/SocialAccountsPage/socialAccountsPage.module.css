.root {
  padding: 2.5rem 3rem;
  max-width: var(--venia-global-maxWidth);
  margin: 0 auto;
}

.heading {
  font-family: var(--venia-global-fontFamily-serif);
  font-weight: var(--venia-global-fontWeight-bold);
  line-height: 1.25em;
  margin-bottom: 2rem;
  text-align: center;
  text-transform: capitalize;
}

.content {
  width: 100%;
  display: grid;
  grid-gap: 3rem;
  grid-template-columns: repeat(2, minmax(300px, 1fr));
}

.accountsHeading {
  margin-bottom: 1rem;
  font-size: 1.375rem;
  font-weight: 300;
  line-height: 1.375rem;
  padding-bottom: 0.5rem;
  border-bottom: 2px solid rgb(var(--venia-global-color-gray));
}

@media (max-width: 1024px) {
  .root {
    padding-left: 1.5rem;
    padding-right: 1.5rem;
  }
}

@media (max-width: 640px) {
  .content {
    grid-template-columns: 1fr;
  }
}
