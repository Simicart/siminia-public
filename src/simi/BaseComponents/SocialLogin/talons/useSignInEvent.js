import SIGNIN_OPERATIONS from '@magento/peregrine/lib/talons/SignIn/signIn.gql.js';
import { GET_CART_DETAILS_QUERY as getCartDetailsQuery } from '@magento/venia-ui/lib/components/SignIn/signIn.gql.js';
import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import { useApolloClient, useMutation } from '@apollo/client';
import { useCallback, useState } from 'react';
import { useCartContext } from '@magento/peregrine/lib/context/cart';
import { useUserContext } from '@magento/peregrine/lib/context/user';
import { useAwaitQuery } from '@magento/peregrine/lib/hooks/useAwaitQuery';
import { clearCartDataFromCache } from '@magento/peregrine/lib/Apollo/clearCartDataFromCache';
import { clearCustomerDataFromCache } from '@magento/peregrine/lib/Apollo/clearCustomerDataFromCache';
import { retrieveCartId } from '@magento/peregrine/lib/store/actions/cart';
import { useEventListener } from '@magento/peregrine';

export const useSignInEvent = (props = {}) => {
  const { handleErrors } = props;
  const operations = mergeOperations(SIGNIN_OPERATIONS, props.operations);

  const {
    createCartMutation,
    getCustomerQuery,
    mergeCartsMutation
  } = operations;

  const apolloClient = useApolloClient();
  const [{ isSignedIn }] = useUserContext();
  const [isSigningIn, setIsSigningIn] = useState(false);

  const [
    { cartId },
    { createCart, removeCart, getCartDetails }
  ] = useCartContext();

  const [{ isGettingDetails }, { getUserDetails, setToken }] = useUserContext();

  const [fetchCartId] = useMutation(createCartMutation);
  const [mergeCarts] = useMutation(mergeCartsMutation);
  const fetchUserDetails = useAwaitQuery(getCustomerQuery);
  const fetchCartDetails = useAwaitQuery(getCartDetailsQuery);

  const socialLoginEventHandler = useCallback(
    async event => {
      setIsSigningIn(true);

      try {
        if (isSignedIn) {
          return null;
        }

        const { token, result, messages, redirect_data } = event.data || {};

        if (result === 1 && token) {
          const sourceCartId = cartId;

          await setToken(token);

          // Clear all cart/customer data from cache and redux.
          await clearCartDataFromCache(apolloClient);
          await clearCustomerDataFromCache(apolloClient);
          await removeCart();
          // Create and get the customer's cart id.
          await createCart({
            fetchCartId
          });
          const destinationCartId = await retrieveCartId();

          // Merge the guest cart into the customer cart.
          await mergeCarts({
            variables: {
              destinationCartId,
              sourceCartId
            }
          });

          // Ensure old stores are updated with any new data.
          getUserDetails({ fetchUserDetails });
          getCartDetails({ fetchCartId, fetchCartDetails });
        } else if (result === 0) {
          handleErrors({
            message: messages[0] || null,
            redirectWithError: redirect_data && redirect_data.redirectWithError
          });
        }

        setIsSigningIn(false);
      } catch (error) {
        if (process.env.NODE_ENV !== 'production') {
          console.error(error);
        }

        setIsSigningIn(false);
      }
    },
    [
      cartId,
      apolloClient,
      removeCart,
      setToken,
      createCart,
      fetchCartId,
      mergeCarts,
      getUserDetails,
      fetchUserDetails,
      getCartDetails,
      fetchCartDetails,
      isSignedIn,
      handleErrors
    ]
  );

  useEventListener(window, 'message', socialLoginEventHandler);

  return {
    isBusy: isGettingDetails || isSigningIn
  };
};
