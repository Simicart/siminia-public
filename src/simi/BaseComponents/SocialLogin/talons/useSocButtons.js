import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useSocialLoginContext } from '../context';
import { useUserContext } from '@magento/peregrine/lib/context/user';

const VISIBLE_BUTTONS_COUNT = 3;

const getPopupParams = () => {
  const bottomPadding = 22;

  const screenX = window.screenX || window.screenLeft;
  const screenY = window.screenY || window.screenTop;
  const outerWidth = window.outerWidth || document.body.clientWidth;
  const outerHeight =
    window.outerHeight || document.body.clientHeight - bottomPadding;

  const width = 500;
  const height = 420;

  const leftDivider = 2;
  const topDivider = 2.5;

  const left = parseInt(screenX + (outerWidth - width) / leftDivider, 10);
  const top = parseInt(screenY + (outerHeight - height) / topDivider, 10);

  return `width=${width},height=${height},left=${left},top=${top}`;
};

export const useSocButtons = (props = {}) => {
  const { buttons, mode, showCreateAccount } = props;
  const isShowMoreBtn =
    mode === 'popup' && buttons.length > VISIBLE_BUTTONS_COUNT;
  const [showAll, setShowAll] = useState(!isShowMoreBtn);
  const { buttonShape, errors, handleErrors } = useSocialLoginContext();
  const [{ isSignedIn, token }] = useUserContext();
  const errorRef = useRef(null);

  const visibleButtons = useMemo(
    () => (showAll ? buttons : buttons.slice(0, VISIBLE_BUTTONS_COUNT)),
    [showAll, buttons]
  );

  const handleLogin = useCallback(
    button => {
      const { type, url } = button;

      handleErrors(null);

      window.open(
        `${url}&isAjax=true${
          !isSignedIn ? '&generateToken=true' : `&token=${token}`
        }`,
        type,
        getPopupParams()
      );
    },
    [isSignedIn, handleErrors, token]
  );

  const handleShowMore = useCallback(() => setShowAll(true), [setShowAll]);

  useEffect(() => {
    const el = errorRef && errorRef.current;

    if (errors) {
      const { redirectWithError } = errors;

      if (redirectWithError && showCreateAccount) {
        handleErrors({
          ...errors,
          redirectWithError: 0
        });

        showCreateAccount();
      }

      if (el && mode === 'popup') {
        el.scrollIntoView();
      }
    }

    return () => {
      if (errors && !errors.redirectWithError) {
        handleErrors(null);
      }
    };
  }, [errors, errorRef, showCreateAccount, handleErrors, mode]);

  return {
    handleLogin,
    visibleButtons,
    isShowMoreBtn: isShowMoreBtn && !showAll,
    handleShowMore,
    buttonShape,
    errors,
    errorRef
  };
};
