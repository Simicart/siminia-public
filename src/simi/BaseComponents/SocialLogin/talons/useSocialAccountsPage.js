import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import DEFAULT_OPERATIONS from './socialLogin.ggl';
import { useLazyQuery, useMutation } from '@apollo/client';
import { useCallback, useEffect, useMemo } from 'react';
import { useUserContext } from '@magento/peregrine/lib/context/user';
import { useHistory } from 'react-router-dom';
import { useSocialLoginContext } from '../context';
import { useEventListener } from '@magento/peregrine';

export const useSocialAccountsPage = (props = {}) => {
  const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
  const { getSocAccountDataQuery, unlinkAccountMutation } = operations;

  const {
    buttons: allButtons,
    handleErrors,
    isEnabled
  } = useSocialLoginContext();
  const [{ isSignedIn }] = useUserContext();
  const history = useHistory();

  const [runQuery, { data, loading }] = useLazyQuery(getSocAccountDataQuery, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
    skip: !isSignedIn || !isEnabled
  });

  const { amSocialLoginAccountData: linkedAccounts } = data || {};

  const [unlinkAccount, { loading: unlinkLoading }] = useMutation(
    unlinkAccountMutation
  );

  const handleUnlink = useCallback(
    async type => {
      try {
        await unlinkAccount({
          variables: { type },
          refetchQueries: [{ query: getSocAccountDataQuery }],
          awaitRefetchQueries: true
        });
      } catch (error) {
        if (process.env.NODE_ENV === 'development') {
          console.error(error);
        }
      }
    },
    [unlinkAccount, getSocAccountDataQuery]
  );

  // If the user is no longer signed in, redirect to the home page.
  useEffect(() => {
    if (!isSignedIn) {
      history.push('/');
    } else {
      runQuery();
    }
  }, [history, isSignedIn, runQuery]);

  const buttons = useMemo(() => {
    if (!allButtons || !linkedAccounts) {
      return null;
    }

    return allButtons.filter(btn =>
      linkedAccounts.every(({ type }) => type !== btn.type)
    );
  }, [linkedAccounts, allButtons]);

  const linkAccountHandler = useCallback(
    event => {
      try {
        if (!isSignedIn) {
          return null;
        }

        const { result, messages, token } = event.data || {};

        if (result === 1 && !token) {
          handleErrors(null);
          runQuery();
        } else if (messages) {
          handleErrors({
            message: messages[0]
          });
        }
      } catch (e) {
        if (process.env.NODE_ENV !== 'production') {
          console.error(error);
        }
      }
    },
    [runQuery, handleErrors, isSignedIn]
  );

  useEventListener(window, 'message', linkAccountHandler);

  return {
    isEnabled,
    linkedAccounts,
    loading,
    handleUnlink,
    buttons,
    isDisabled: unlinkLoading
  };
};
