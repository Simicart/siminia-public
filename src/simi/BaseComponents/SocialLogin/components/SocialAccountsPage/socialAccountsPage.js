import React from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { shape, string } from 'prop-types';
import { FormattedMessage, useIntl } from 'react-intl';
import defaultClasses from './socialAccountsPage.module.css';
import { useSocialAccountsPage } from '../../talons/useSocialAccountsPage';
import { Title } from '@magento/venia-ui/lib/components/Head';
import { fullPageLoadingIndicator } from '@magento/venia-ui/lib/components/LoadingIndicator';
import LinkedAccounts from './linkedAccounts';
import SocialButtons from '../SocialButtons';
import ErrorView from '@magento/venia-ui/lib/components/ErrorView';

const SocialAccountsPage = props => {
  const {
    linkedAccounts,
    loading,
    handleUnlink,
    buttons,
    isDisabled,
    isEnabled
  } = useSocialAccountsPage();

  const { formatMessage } = useIntl();

  if (!isEnabled) {
    return (
      <ErrorView
        message={formatMessage({
          id: 'magentoRoute.routeError',
          defaultMessage:
            "Looks like the page you were hoping to find doesn't exist. Sorry about that."
        })}
      />
    );
  }

  const classes = mergeClasses(defaultClasses, props.classes);

  const title = formatMessage({
    id: 'amSocialLogin.socialAccountsTitle',
    defaultMessage: 'My Social Accounts'
  });

  if (loading) {
    return fullPageLoadingIndicator;
  }

  return (
    <div className={classes.root}>
      <Title>{title}</Title>
      <h1 className={classes.heading}>{title}</h1>

      <div className={classes.content}>
        <LinkedAccounts
          accounts={linkedAccounts}
          isDisabled={isDisabled}
          handleUnlink={handleUnlink}
        />

        <div className={classes.accountsBlock}>
          <div className={classes.accountsHeading}>
            <FormattedMessage
              id={'amSocialLogin.linkNewAccounts'}
              defaultMessage={'Link new accounts'}
            />
          </div>
          <SocialButtons buttons={buttons} mode={'account'} />
        </div>
      </div>
    </div>
  );
};

SocialAccountsPage.propTypes = {
  classes: shape({
    root: string
  })
};

export default SocialAccountsPage;
