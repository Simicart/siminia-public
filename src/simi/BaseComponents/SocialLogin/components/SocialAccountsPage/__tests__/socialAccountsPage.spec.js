import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import SocialAccountsPage from '../socialAccountsPage';
import { useSocialAccountsPage } from '../../../talons/useSocialAccountsPage';

jest.mock('../linkedAccounts', () => 'LinkedAccounts');
jest.mock('../../SocialButtons', () => 'SocialButtons');
jest.mock('@magento/venia-ui/lib/components/Head', () => ({
  Title: () => 'Title'
}));
jest.mock('../../../talons/useSocialAccountsPage', () => {
  const useSocialAccountsPage = jest.fn(() => {});

  return { useSocialAccountsPage };
});
jest.mock('@magento/venia-ui/lib/components/ErrorView', () => 'ErrorView');
jest.mock('react-router-dom', () => {
  return {
    useHistory: jest.fn(() => ({ push: jest.fn() }))
  };
});
jest.mock('../../../context', () => {
  const useSocialLoginContext = jest.fn(() => {});

  return { useSocialLoginContext };
});

jest.mock('react-intl', () => ({
  FormattedMessage: props => (
    <div componentName="Formatted Message Component" {...props} />
  ),
  useIntl: jest.fn().mockReturnValue({
    formatMessage: jest
      .fn()
      .mockImplementation(options => options.defaultMessage)
  })
}));

jest.mock('@magento/venia-ui/lib/components/LoadingIndicator', () => {
  return {
    fullPageLoadingIndicator: 'LoadingIndicator'
  };
});

test('render errorView', () => {
  useSocialAccountsPage.mockReturnValue({ isEnabled: false, buttons: null });

  const tree = createTestInstance(<SocialAccountsPage />);
  expect(tree.toJSON()).toMatchSnapshot();
});

test('render loading', () => {
  useSocialAccountsPage.mockReturnValue({ isEnabled: true, loading: true });

  const tree = createTestInstance(<SocialAccountsPage />);
  expect(tree.toJSON()).toMatchSnapshot();
});

test('render correct page', () => {
  useSocialAccountsPage.mockReturnValue({
    isEnabled: true,
    loading: false,
    buttons: [
      {
        label: 'Facebook',
        type: 'facebook'
      },
      {
        label: 'Instagram',
        type: 'instagram'
      }
    ]
  });

  const tree = createTestInstance(<SocialAccountsPage />);
  expect(tree.toJSON()).toMatchSnapshot();
});
