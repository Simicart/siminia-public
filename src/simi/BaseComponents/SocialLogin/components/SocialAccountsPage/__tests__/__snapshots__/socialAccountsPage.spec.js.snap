// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correct page 1`] = `
<div>
  Title
  <h1>
    My Social Accounts
  </h1>
  <div>
    <LinkedAccounts />
    <div>
      <div>
        <div
          componentName="Formatted Message Component"
          defaultMessage="Link new accounts"
          id="amSocialLogin.linkNewAccounts"
        />
      </div>
      <SocialButtons
        buttons={
          Array [
            Object {
              "label": "Facebook",
              "type": "facebook",
            },
            Object {
              "label": "Instagram",
              "type": "instagram",
            },
          ]
        }
        mode="account"
      />
    </div>
  </div>
</div>
`;

exports[`render errorView 1`] = `
<ErrorView
  message="Looks like the page you were hoping to find doesn't exist. Sorry about that."
/>
`;

exports[`render loading 1`] = `"LoadingIndicator"`;

exports[`render page without linked accounts 1`] = `
<div>
  Title
  <h1>
    My Social Accounts
  </h1>
  <div>
    <LinkedAccounts />
    <div>
      <div>
        <div
          componentName="Formatted Message Component"
          defaultMessage="Link new accounts"
          id="amSocialLogin.linkNewAccounts"
        />
      </div>
      <SocialButtons
        buttons={
          Array [
            Object {
              "label": "Facebook",
              "type": "facebook",
            },
            Object {
              "label": "Instagram",
              "type": "instagram",
            },
          ]
        }
        mode="account"
      />
    </div>
  </div>
</div>
`;
